sidecar/src/agent_remote_config.rs

```diff
@@ -0,0 +1,212 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present Datadog, Inc.
+
+use datadog_ipc::platform::{FileBackedHandle, MappedMem, NamedShmHandle, ShmHandle};
+use ddcommon::Endpoint;
+use std::hash::{Hash, Hasher};
+use std::io;
+use std::io::Write;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Mutex;
+use zwohash::ZwoHasher;
+
+pub struct AgentRemoteConfigWriter<T>
+where
+    T: FileBackedHandle<T> + From<MappedMem<T>>,
+{
+    handle: Mutex<Option<MappedMem<T>>>,
+}
+
+pub struct AgentRemoteConfigReader<T>
+where
+    T: FileBackedHandle<T> + From<MappedMem<T>>,
+{
+    handle: Option<MappedMem<T>>,
+    endpoint: Option<Endpoint>,
+    current_config: Option<Vec<u8>>,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+struct RawMetaData {
+    generation: u64,
+    size: usize,
+    writing: AtomicBool,
+}
+
+#[repr(C)]
+#[derive(Debug)]
+struct RawData {
+    meta: RawMetaData,
+    buf: [u8],
+}
+
+impl RawData {
+    fn as_slice(&self) -> &[u8] {
+        // Safety: size is expected to be truthful
+        unsafe { std::slice::from_raw_parts(self.buf.as_ptr(), self.meta.size) }
+    }
+
+    fn as_slice_mut(&mut self) -> &mut [u8] {
+        // Safety: size is expected to be truthful
+        unsafe { std::slice::from_raw_parts_mut(self.buf.as_mut_ptr(), self.meta.size) }
+    }
+}
+
+impl From<&[u8]> for &RawData {
+    fn from(value: &[u8]) -> Self {
+        // Safety: MappedMem is supposed to be big enough
+        unsafe { &*(value as *const [u8] as *const RawData) }
+    }
+}
```

Comment:
I disagree with the safety of this operation.
First, in practice the slice is not guaranteed to have the right alignement, since you're copying data into a Vec, and then casting the underlying data to `&RawData` which can crash on ARM.
Second since this is a general trait, this can theoretically be called on any slice so the Safety comment is not really exact.

Reply:
In practice the Slice is mapped to the start of a page, so it should always be properly aligned, on a page of size >+ 4096.

Reply:
You copy the page data to a Vec<u8> in the reader before casting it 

https://github.com/DataDog/libdatadog/blob/2cbbe1ee3318d00513c9e7bbc79180805f976c59/sidecar/src/agent_remote_config.rs#L143-L148

Reply:
So, what do you recommend here? I haven't seen any crashes any my M1 yet, but yeah... Is it possible to allocate an aligned Vec?

Reply:
Does it make sense to allocate a Vec<u64> instead? RawMetaData is repr(C), hence 64 bit aligned.
This should then make it safe, I assume?
And then reinterpret the &[u64] as &[u8] for returning.

Reply:
I thin that instead of copying and keeping the whole memory in a contiguous slice, you can clone the sized RawMetaData to a field, and only the buf to a Vec.
Then you don't have to cast the whole vec to RawMetadata